src/RMGOutputTools.cc

```diff
@@ -0,0 +1,351 @@
+          second_input_hits.front()->energy_deposition += hit->energy_deposition;
+          hit->energy_deposition = 0;
+        }
+      }
+    }
+  }
+}
+
+
+std::map<int, std::vector<RMGDetectorHit*>> RMGOutputTools::
+    combine_low_energy_tracks(std::map<int, std::vector<RMGDetectorHit*>> hits_map,
+        RMGOutputTools::ClusterPars cluster_pars, bool has_distance_to_surface) {
+
+  RMGLog::Out(RMGLog::debug, "Merging low energy electron tracks ");
+
+
+  // for tracks below an energy threshold look for a close neighbour to merge it with
+  // only done for e-.
+  std::map<int, std::vector<RMGDetectorHit*>> output_hits = hits_map;
+
+  for (const auto& [trackid, input_hits] : hits_map) {
+
+    // only apply to e-
+    if (input_hits.front()->particle_type != 11) continue;
+
+    // compute energy of each track
+    double energy = 0;
+    for (auto hit : input_hits) { energy += hit->energy_deposition; }
+
+    // continue for high energy tracks
+    if (energy > cluster_pars.track_energy_threshold) continue;
+
+    // distance threshold to merge tracks
+    double threshold =
+        (not has_distance_to_surface) or
+                ((input_hits.front()->distance_to_surface_prestep) > cluster_pars.surface_thickness)
+            ? cluster_pars.cluster_distance
+            : cluster_pars.cluster_distance_surface;
+
+    // now search for another track to merge it with
+    int cluster_trackid = -1;
+    float cluster_energy = 0;
+
+    // loop over secondary tracks
+    for (const auto& [second_trackid, second_input_hits] : hits_map) {
+      if (second_trackid == trackid) continue;
+
+      // only cluster into high energy tracks
+      energy = 0;
+      for (auto hit : second_input_hits) { energy += hit->energy_deposition; }
+
+
+      // compute distance between the first step of this track and that of the
+      // other track.
+      if (energy > cluster_energy && (input_hits.front()->global_position_prestep -
+                                         second_input_hits.front()->global_position_prestep)
+                                             .mag() < threshold) {
+
+        cluster_energy = energy;
+        cluster_trackid = second_trackid;
+      }
+    }
+
+    if (cluster_trackid != -1) {
+
+      // change all the track-ids
+      for (auto hit : output_hits[trackid]) { hit->track_id = cluster_trackid; }
+
+      // add these elements to the start of the second track
+      output_hits[cluster_trackid].insert(output_hits[cluster_trackid].begin(),
+          output_hits[trackid].begin(), output_hits[trackid].end());
+
+      output_hits.erase(trackid);
+      RMGLog::Out(RMGLog::debug, "Removing trackid ", trackid);
+    }
+  }
+  return output_hits;
+}
+
+RMGDetectorHitsCollection* RMGOutputTools::pre_cluster_hits(const RMGDetectorHitsCollection* hits,
+    RMGOutputTools::ClusterPars cluster_pars, bool has_distance_to_surface, bool has_velocity) {
+
+  // organise hits into a map based on trackid
```

Comment:
do we need a short-cut when `hits` is empty here? this is not checked in the output schemes before calling here.